enhancements/openshift-route-admission-policy.md

```diff
@@ -0,0 +1,133 @@
+
+## Summary
+
+See the [section][#route-admission-policies]
+
+## Motivation
+
+See the [section][#route-admission-policies]
+
+### Goals
+
+1. Define an API that allows the cluster administrator to enable sharing of hostnames across namespaces.
+
+### Non-Goals
+
+## Proposal
+
+Modify the existing `operator.openshift.io`'s custom resource [`IngressController`][ingress-controller-api] to include an enum field `RouteAdmissionPolicy` that can be set to `AllowInterNamespaceClaims` or `Strict`, where `Strict` is default.
+
+
+### User Stories
+
+#### Default Route Admission Policy
+
+The default behavior is `Strict` which match the current behavior and is a secure default.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+The openshift-ingress-operator will be able to use the enum `RouteAdmissionPolicy` to enable the environment variable `ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK` in the openshift-ingress router. This is enabled when `RouteAdmissionPolicy` is set to `AllowInterNamespaceClaims`.
+
+### Risks and Mitigations
+
+Addition of the new field requires no migration, as the default matches the current configuration.
+
+## Design Details
+
+### New API
+
+`RouteAdmissionPolicy` enum definition.
+
+```go
+// RouteAdmissionPolicy is an admission policy for allowing new route claims.
+type RouteAdmissionPolicy string
+
+const (
+    // AllowInterNamespaceClaimsRouteAdmissionPolicy allows routes to claim the same hostname across namespaces.
+    AllowInterNamespaceClaimsRouteAdmissionPolicy RouteAdmissionPolicy = "AllowInterNamespaceClaims"
+
+    // StrictRouteAdmissionPolicy does not allow routes to claim the same hostname across namespaces.
+    // This is the default behavior 
+    StrictRouteAdmissionPolicy RouteAdmissionPolicy = "Strict"
+)
+```
+
+### Migration of existing clusters
+
+All clusters with `empty` RouteAdmissionPolicy can continue running with the current default, without setting a value.
+
+#### Rollbacks
+
+In case of rollbacks where AllowInterNamespaceClaims has been used, administrators would have to move routes to a shared namespace.
+
+### Test Plan
+
+- Create two routes in two different namespaces with the same hostname, and ensure that they are not admitted by default (`empty` or `unset`) and `Strict`.
+
+- Create two routes in two different namespaces with the same hostname, and ensure that they are admitted when RouteAdmissionPolicy is set to `AllowInterNamespaceClaims`.
+
+#### Migration
+
+No migration should be required.
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
+
+See the [rollback section][#rollbacks]
+
+### Version Skew Strategy
+
+The API defines `empty` equivalent to `Strict`, therefore all the clients will have to keep that check to remain backwards compatible.
+
+## Implementation History
+
+## Drawbacks
+
+This only implements one existing feature in the OpenShift router that existed in OpenShift 3.x, whereas there are many more configurations.
```

Comment:
The question is how to re-introduce these configurables in an organized way.  Currently the proposal defines a new enum but does not say how it is used—would the enum be the type of a new field that would be added directly to `IngressControllerSpec`, or should we introduce a new type for this and other configurables (such as `ROUTER_ALLOW_WILDCARD_ROUTES` and allowed/denied domains) and add a field of that new type to `IngressControllerSpec`?

Reply:
Good question, do you know what other configurables are most often used "in the wild" in OpenShift 3.x that should be supported, I assume OpenShift 4 should be simplified and more opinionated?

Reply:
It's a useful conversation, but is this on topic for the PR?

Reply:
The discussion of which openshift-router configurables need to be supported in OpenShift 4 goes beyond the scope of this PR, but I believe it would be useful to consider `ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK` in the context of other configurables that could be introduced in the future.

Let's assume that we want to support `ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK` and, as an example, one or more other configurables that pertain to how openshift-router admits routes (`ROUTER_UPDATE_STATUS`, `ROUTER_OVERRIDE_DOMAINS`, `ROUTER_DENIED_DOMAINS`, `ROUTER_ALLOWED_DOMAINS`, `ROUTER_ALLOW_WILDCARD_ROUTES`, `EXTENDED_VALIDATION`); would it be reasonable to define a `RouteAdmissionPolicy` type with fields for `NamespaceOwnershipCheck` (or whatever, using the proposed enum) and other configurables related to route admission?

To be clear, this enhancement proposal can be restricted in scope to introducing `ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK`, but it should introduce it in a way that establishes a pattern for organized API growth should we (re-)introduce other configurables in the future.

Reply:
Addressed here: https://github.com/openshift/enhancements/pull/155